orc/src/main/java/org/apache/iceberg/orc/OrcMetrics.java

```diff
@@ -19,51 +19,211 @@
+        upperBounds);
+  }
+
   static Metrics fromWriter(Writer writer) {
-    // TODO: implement rest of the methods for ORC metrics in
-    // https://github.com/apache/incubator-iceberg/pull/199
-    return new Metrics(writer.getNumberOfRows(),
-        null,
-        null,
-        Collections.emptyMap(),
-        null,
-        null);
+    try {
+      return buildOrcMetrics(writer.getNumberOfRows(),
+          writer.getSchema(), writer.getStatistics());
+    } catch (IOException ioe) {
+      throw new RuntimeIOException(ioe, "Failed to get statistics from writer");
+    }
+  }
+
+  private static long toMicros(Timestamp ts) {
+    return ts.getTime() * 1000;
+  }
+
+  private static Optional<ByteBuffer> fromOrcMin(Types.NestedField column,
+                                                 ColumnStatistics columnStats) {
+    ByteBuffer min = null;
+    if (columnStats instanceof IntegerColumnStatistics) {
+      IntegerColumnStatistics intColStats = (IntegerColumnStatistics) columnStats;
+      min = column.type().typeId() == Type.TypeID.INTEGER ?
+          Conversions.toByteBuffer(column.type(), (int) intColStats.getMinimum()) :
+          Conversions.toByteBuffer(column.type(), intColStats.getMinimum());
+    } else if (columnStats instanceof DoubleColumnStatistics) {
+      double minVal = ((DoubleColumnStatistics) columnStats).getMinimum();
+      min = column.type().typeId() == Type.TypeID.DOUBLE ?
+          Conversions.toByteBuffer(column.type(), minVal) :
+          Conversions.toByteBuffer(column.type(), (float) minVal);
+    } else if (columnStats instanceof StringColumnStatistics) {
+      min = Conversions.toByteBuffer(column.type(),
+          ((StringColumnStatistics) columnStats).getMinimum());
+    } else if (columnStats instanceof DecimalColumnStatistics) {
+      min = Optional
+          .ofNullable(((DecimalColumnStatistics) columnStats).getMinimum())
+          .map(minStats -> {
+            BigDecimal minValue = minStats.bigDecimalValue()
+                .setScale(((Types.DecimalType) column.type()).scale());
+            return Conversions.toByteBuffer(column.type(), minValue);
+          })
+          .orElse(null);
+    } else if (columnStats instanceof DateColumnStatistics) {
+      min = Optional.ofNullable(((DateColumnStatistics) columnStats).getMinimum())
+          .map(minStats -> Conversions.toByteBuffer(column.type(),
+              (int) ChronoUnit.DAYS.between(EPOCH_DAY,
+                  EPOCH.plus(minStats.getTime(), ChronoUnit.MILLIS).toLocalDate())))
+          .orElse(null);
+    } else if (columnStats instanceof TimestampColumnStatistics) {
+      TimestampColumnStatistics tColStats = (TimestampColumnStatistics) columnStats;
+      Timestamp minValue = ((Types.TimestampType) column.type()).shouldAdjustToUTC() ?
+          tColStats.getMinimum() : tColStats.getMinimumUTC();
+      min = Optional.ofNullable(minValue)
+          .map(v -> Conversions.toByteBuffer(column.type(), toMicros(v)))
```

Comment:
I don't think this `toMicros` is correct. It just multiplies a milliseconds value by 1,000 without adjusting for the nanoseconds component of the timestamp.

Reply:
As per the spec, `TimestampStatistics` only stores stats with millisecond precision so that's why there's no adjustment for sub-millisecond. As mentioned in a comment below by @shardulm94 we could opt to not use timestamp stats for the moment until [ORC-611](https://issues.apache.org/jira/browse/ORC-611) is addressed.

Reply:
I think it would be better to increment the upper bound value by 1ms to account for ORC-611, instead of not writing the bounds for timestamps. Iceberg doesn't require the bound to be exact, and the file-level filtering enabled by this is really valuable for cutting down on data that needs to be processed.